File: src/black_scholes.py
import scipy.stats as stats
import math

def CallValue(S: float, K: float, sigma: float, t: float, r: float) -> float:
    """
    Berechnet den Preis einer Call-Option nach dem Black-Scholes-Modell.

    Parameter:
    -----------
    S: float
        Aktueller Aktienkurs (Current Stock Price)
    K: float
        Ausübungspreis der Option (Strike Price)
    sigma: float
        Volatilität der Aktie (annualisierte Standardabweichung der Renditen)
    t: float
        Zeit bis zur Fälligkeit in Tagen (Days to Maturity)
    r: float
        Risikofreier Zinssatz (annualisiert, z.B. 0.05 für 5%)

    Rückgabewert:
    -------------
    float: Preis der Call-Option
    """
    t1 = t / 365 + 0.00000001  # Umrechnung der Tage in Jahre + kleiner Puffer, um Division durch Null zu vermeiden
    d1 = (math.log(S / K) + t1 * (r + sigma ** 2 / 2)) / (sigma * math.sqrt(t1))
    d2 = d1 - sigma * math.sqrt(t1)
    c1 = S * stats.norm.cdf(d1, 0, 1)  # S * N(d1)
    c2 = K * math.exp(-r * t1) * stats.norm.cdf(d2, 0, 1)  # K * e^(-r*t) * N(d2)
    CallValue = c1 - c2
    return CallValue

def PutValue(S: float, K: float, sigma: float, t: float, r: float) -> float:
    """
    Berechnet den Preis einer Put-Option nach dem Black-Scholes-Modell unter Verwendung der Put-Call-Parität.

    Parameter:
    -----------
    S: float
        Aktueller Aktienkurs (Current Stock Price)
    K: float
        Ausübungspreis der Option (Strike Price)
    sigma: float
        Volatilität der Aktie (annualisierte Standardabweichung der Renditen)
    t: float
        Zeit bis zur Fälligkeit in Tagen (Days to Maturity)
    r: float
        Risikofreier Zinssatz (annualisiert, z.B. 0.05 für 5%)

    Rückgabewert:
    -------------
    float: Preis der Put-Option
    """
    cv = CallValue(S, K, sigma, t, r)  # Preis der Call-Option
    t2 = t / 365 + 0.00000001  # Umrechnung der Tage in Jahre + kleiner Puffer
    PutValue = cv - S + K * math.exp(-r * t2)  # Put-Call-Parität: Put = Call - S + K * e^(-r*t)
    return PutValue

def ProbLessThan(x: float, S: float, IV: float, t: float, r: float) -> float:
    """
    Berechnet die Wahrscheinlichkeit, dass der Aktienkurs unter einem bestimmten Wert x liegt,
    basierend auf dem Black-Scholes-Modell.

    Parameter:
    -----------
    x: float
        Zielwert des Aktienkurses (z.B. Strike Price)
    S: float
        Aktueller Aktienkurs (Current Stock Price)
    IV: float
        Implizite Volatilität (Implied Volatility)
    t: float
        Zeit bis zur Fälligkeit in Tagen (Days to Maturity)
    r: float
        Risikofreier Zinssatz (annualisiert, z.B. 0.05 für 5%)

    Rückgabewert:
    -------------
    float: Wahrscheinlichkeit, dass der Aktienkurs unter x liegt
    """
    if x == 0:
        x = 0.0001  # Vermeidet Division durch Null oder numerische Probleme
    prob_less_than = stats.norm.cdf((math.log(x / S) - r * (t / 365)) / (IV * math.sqrt(t / 365)))
    return prob_less_than

File: frontend/src/lib/blackScholes.ts
/**
 * Black-Scholes-Optionsbewertung — Portierung von `src/black_scholes.py`.
 * Wird client-seitig im Risikograf für die IV-Shift-Neubewertung genutzt.
 *
 * `t` ist in TAGEN (wie im Python-Original), intern → Jahre.
 */

/** Standardnormal-CDF via Abramowitz-Stegun 7.1.26 (Ersatz für scipy norm.cdf). */
export function normCdf(x: number): number {
  const t = 1 / (1 + 0.2316419 * Math.abs(x));
  const d = 0.3989422804014327 * Math.exp(-(x * x) / 2); // 1/sqrt(2π) * e^(-x²/2)
  let p =
    d *
    t *
    (0.319381530 +
      t * (-0.356563782 + t * (1.781477937 + t * (-1.821255978 + t * 1.330274429))));
  p = 1 - p;
  return x >= 0 ? p : 1 - p;
}

/** Call-Preis nach Black-Scholes. S=Kurs, K=Strike, sigma=IV, t=Tage, r=Zins. */
export function callValue(S: number, K: number, sigma: number, t: number, r: number): number {
  const t1 = t / 365 + 1e-8;
  const sqrtT = sigma * Math.sqrt(t1);
  if (sqrtT === 0 || S <= 0 || K <= 0) return Math.max(S - K, 0);
  const d1 = (Math.log(S / K) + t1 * (r + (sigma * sigma) / 2)) / sqrtT;
  const d2 = d1 - sqrtT;
  return S * normCdf(d1) - K * Math.exp(-r * t1) * normCdf(d2);
}

/** Put-Preis via Put-Call-Parität (wie im Python-Original). */
export function putValue(S: number, K: number, sigma: number, t: number, r: number): number {
  const cv = callValue(S, K, sigma, t, r);
  const t2 = t / 365 + 1e-8;
  return cv - S + K * Math.exp(-r * t2);
}
